src/presign/participant.rs

```diff
@@ -85,6 +85,37 @@ pub enum Status {
     TerminatedSuccessfully,
 }
 
+/// PresignContext includes shared parameters in
+/// [`SharedContext`] and public parameters as
+/// defined in [`AuxInfoPublic`].
+pub struct PresignContext {
+    shared_context: SharedContext,
+    auxinfo_public: Vec<AuxInfoPublic>,
+}
+
+impl ProofContext for PresignContext {
+    fn as_bytes(&self) -> Result<Vec<u8>> {
+        Ok([
+            self.shared_context.as_bytes()?,
+            bincode::serialize(&self.auxinfo_public)
+                .map_err(|_| InternalError::InternalInvariantFailed)?,
+        ]
+        .concat())
+    }
+}
+
+impl PresignContext {
+    pub(crate) fn collect<P: InnerProtocolParticipant>(
+        auxinfo_public: &[AuxInfoPublic],
+        p: &P,
+    ) -> Self {
+        Self {
+            shared_context: SharedContext::collect(p),
+            auxinfo_public: auxinfo_public.to_vec(),
```

Comment:
I don't mind it!

Reply:
That said, I think it'd be nice to put all the collecting and sorting in one place, instead of having it spread across this function and `retrieve_context` below. Since this already takes a participant, you can take the retrieve-and-sort lines from below and put them into this function, and then you can pass the `Vec` that you sorted directly.